datafusion/core/tests/sqllogictests/test_files/array.slt

```diff
@@ -907,14 +907,122 @@ select array_positions(column1, make_array(4, 5, 6)), array_positions(make_array
 [6] []
 [1] []
 
-## array_replace
+## array_replace (replaces one occurrence of an element) (aliases: `list_replace`)
 
-# array_replace scalar function
+# array_replace scalar function #1
 query ???
 select array_replace(make_array(1, 2, 3, 4), 2, 3), array_replace(make_array(1, 4, 4, 5, 4, 6, 7), 4, 0), array_replace(make_array(1, 2, 3), 4, 0);
 ----
+[1, 3, 3, 4] [1, 0, 4, 5, 4, 6, 7] [1, 2, 3]
+
+# array_replace scalar function #2 (with optional argument)
+query ???
+select array_replace(make_array(1, 2, 3, 4), 2, 3, 2), array_replace(make_array(1, 4, 4, 5, 4, 6, 7), 4, 0, 2), array_replace(make_array(1, 2, 3), 4, 0, 3);
```

Comment:
> What do you think of these features, @alamb and @jayzhan211? Should they be implemented or not?


I found the fact that `array_replace` has 4 arguments very confusing initially. All "replace" APIs I can think of (e.g. Rust's [`replace`](https://doc.rust-lang.org/std/primitive.str.html#method.replace)) take only three arguments: the value, the pattern to search for and the replacement value. 

If we want to implement "replace the first N" values, I think we should add a new function like `array_replace_n` (modeled after [`replacen`](https://doc.rust-lang.org/std/primitive.str.html#method.replacen)) 

It would be great to share the implementation as you have done in this PR